### ISSUE TITLE
Investigate moving OS specific Socket and Thread to the Patch Page

### ISSUE THREAD

**Issue by noloader** [weidai11/cryptopp#208]
The Crypto++ wiki has a [Patch Page](http://www.cryptopp.com/wiki/Category:Patch). Its for components that may be useful to others, but don't quite fit into the library proper. There's also a license requirement for inclusion on the patch page, but its not relevant for this discussion.

Denis Bider offered some arguments to remove some code from the library at [issue 23: Recent Windows 8 and Windows 10 support broke Windows 7/VS 2012](https://github.com/weidai11/cryptopp/issues/178#issuecomment-227908380). If I am parsing the comment correctly, DB suggests removing `Socket` and `Pipe` from the library.

I'd like to use this bug report to capture feedback, and track which files and classes are potential targets.

**Comment by zabulus** [comment]
Fully agree. It makes much questions about what pipes, threads and so on are doing in crypto library.

**Comment by zabulus** [comment]
Also cutting these parts will simplify further porting of stuff

**Comment by DevJPM** [comment]
TL;DR: I'm pro-removal.

A quick search showed that it was added in 4.0 and there's also a mention of sinks and sources in 5.0 related to sockets, so it's obviously only in the library so you can stream from a socket or a pipe into Crypto++'s filter system (decrypt, verify, decompress, decode it) and then have it at hand. So the original intent was to not limit people to files but also allow them to use advanced IPC and network communication methods with this fancy mechanism of filters, sources and sinks.

So far for the history lesson. Nowadays we don't need this any longer at network level. We have Boost.ASIO (or your favorite networking library) and we have libraries like OpenSSL, LibreSSL, BoringSSL, GnuTLS, s2n, cryptlib,... which can all handle TLS (a really good protocol for most use-cases) so I don't think we should encourage people here to "roll their own". A

**Comment by noloader** [comment]
Also see [Remove Thread and Socket classes](https://groups.google.com/d/msg/cryptopp-users/5btwLoxXXD4/C_lB6jFpDAAJ) on the mailing list.
